src/main/webapp/js/profilePage.js

```diff
@@ -1,13 +1,79 @@
+* Sets up the functionality of the username input box and the submit button.
 */
-async function getUserName() {
+async function loadProfilePage() {
+  let usernameBox = document.getElementById('userName');
+  let timer; // current ms since last keyup
+  let waitTime = 500; // wait 500ms after last keyup before we ask the server if this username is taken
+
   await fetch('/load-authentication-data').then(response => response.json()).then(async (data) => {
     if (!data.loggedIn) {
       window.location.replace(data.loginUrl);
     }
   });  
   fetch('/load-currentuser-data').then(response => response.json()).then(async (data) => {
-    document.getElementById('userName').value = data.username;
+    usernameBox.value = data.username;
   });
+
+  usernameBox.addEventListener("keyup", () => {
+    if(event.keyCode === 13) { // enter button
+      document.getElementById('submit-username-button').click();
+    } else {
+      clearTimeout(timer);
+      timer = setTimeout(displayAvailability, waitTime);
+    }
+  });
+}
+
+async function isTaken(desiredUsername) {
+  const params = new URLSearchParams();
+  params.append('username', desiredUsername);
+
+  let result;
+  await fetch('/username-available-data', {method: 'post', body: params}).then(response => response.json()).then(taken => {
+    result = taken;
+  });
+  return result;
+}
+
+async function getAvailabilityText(desiredUsername) {
+  if(desiredUsername.length == 0) {
+    return 'Username must be at least 1 character'
+  }
+  if(desiredUsername.length > 20) {
+    return 'Username cannot be more than 20 characters'
+  }
+  let pattern = new RegExp(/^[ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz._0123456789]*$/);
+  if(!pattern.test(desiredUsername)) {
+    return 'Only letters, digits, underscores, and periods are allowed';
+  }
+  if(await isTaken(desiredUsername)) {
+    return 'Username is already taken'
+  }
+  return 'Available';
+}
+
+async function displayAvailability() {
+  desiredUsername = document.getElementById('userName').value;
+  let availabilityBox = document.getElementById('username-availability-message');
+  availabilityBox.innerText = await getAvailabilityText(desiredUsername);
+  if(availabilityBox.innerText == 'Available') {
+      availabilityBox.className = 'green-text';
+  } else {
+      availabilityBox.className = 'red-text';
+  }
+}
+
+async function submitUsername() {
+  desiredUsername = document.getElementById('userName').value;
+  let availabilityBox = document.getElementById('username-availability-message');
+  availabilityBox.innerText = await getAvailabilityText(desiredUsername);
+  if(availabilityBox.innerText == 'Available') {
+    const params = new URLSearchParams();
+    params.append('userName', desiredUsername);
+    await fetch('/create-username-data', {method: 'post', body: params});
+    window.location = '/profilePage.html'
+  } else {
+    alert(availabilityBox.innerText);
```

Comment:
Does this have to be an alert, is there a better way to do this for UX?

Reply:
I think an alert for this looks pretty decent. If you have a better idea please let me know.

Reply:
Maybe show it to me tomorrow over GVC? Also, I noticed that when you delete characters it doesn't recheck if it's good or not, is that something we can fix? So for instance, if I type saraM it says available but if I put a backspace to Sara it doesn't say its unavailable even though it isn't available